### ISSUE TITLE
Can't write to MMIO Registers

### ISSUE THREAD

**Issue by Kevingislason** [RusPiRo/ruspiro-register#8]
Hi RusPiRo maintainers. It seems I'm having trouble writing to MMIO registers on my Raspi at all. Here is a reduced case based on one of the examples in your readme. I simply try to write to and then read from a register.

```
#[macro_use]
extern crate ruspiro_boot;
extern crate ruspiro_allocator;
use ruspiro_register::*;

come_alive_with!(kernel_alive);
run_with!(kernel_run);

define_mmio_register!(RW_REGISTER<ReadWrite<u32>@(0xFF00_0008)>);

pub fn kernel_alive(core: u32) {
    println!("register value is {}", RW_REGISTER::Register.get());  //expect 0
    RW_REGISTER::Register.set(0xFFF); // update a specific field value of the register
    println!("register value is {}", RW_REGISTER::Register.get());  //expect 0xFFF
}

pub fn kernel_run(core: u32) -> ! {
    println!("Kernel running on core {}", core);
    loop {}
}
```

When I power up my raspi, I see this in my console--the register's value doesn't change after the write:
```
########## RusPiRo ----- Bootstrapper v0.3 @ Aarch32 ----- ##########
register value is 0
register value is 0
Kernel running on core 0
```

Any idea what's up here? I'm pretty green at embedded programming, so forgive me if I've made some obvious mistake. Thanks for your help.

**Comment by 2ndTaleStudio** [comment]
Hi Kevingislason,

thanks for giving RusPiRo a try :)

Could you share which Raspberry Pi Model you are using and whether you test this on real hardware or in QEMU?

The memory address you are using ‘0xFF00_0008’ seems a bit arbitrary to me. This might not even be a valid memory address based on your model. On Raspberry Pi the MMIO address range is used to access and configure built in peripheral devices like the GPIO’s and starts at 0x3F00_0000 on a RPi 3 and 0xFE00_0000 on a RPi 4.

Assuming this is the correct MMIO address you are using the behavior of writing to and reading from this address depend on the peripheral that is accessed here. Some peripherals have write only MMIO registers that always read back as zero. A good next step could be to check the documentation of the peripheral you are about to access and configure to see the behavior of the MMIO register you are dealing with.

In a nutshell, MMIO registers are typically bound to a specific peripheral device and not always working as general purpose registers would do.

BR
Schnoogle

**Comment by Kevingislason** [comment]
Thanks for you response! 

> Could you share which Raspberry Pi Model you are using and whether you test this on real hardware or in QEMU?

I'm using a physical Raspi 2B.

>The memory address you are using ‘0xFF00_0008’ seems a bit arbitrary to me

Yeah, ‘0xFF00_0008’ has no special significance for me. I just thought it might be an appropriate test case since I saw it in the readme.

My actual use case involves accessing my Raspi's hardware random number generator, which begins at '0x7e104000' (documented [here](https://elinux.org/BCM2835_registers#RNG)). I haven't been able to write to any of the RNG's registers either.

**Comment by 2ndTaleStudio** [comment]
Hi,

well I see, using those arbitrary address in the README has caused the confusion. Well for the Raspberry 2 the MMIO base address is also 0x3F00_0000. The address range you see in those documents reflect the GPU internal address range of RPi. This address is not accessible from RPi CPU. So you need to replace 0x7Exx_xxxx always with 0x3Fxx_xxxxx within your RPi MMIO register definitions. Than it should work.

So instead of `0x7e10_4000` you may use `0x3f10_4000`.

**Comment by Kevingislason** [comment]
> So instead of 0x7e10_4000 you may use 0x3f10_4000

Seems to have solved my problem!

> I'm pretty green at embedded programming, so forgive me if I've made some obvious mistake

Looks like this^^ was accurate 😅. Sincere thanks for your help
